### ISSUE TITLE
Mutable state & closures on the UI thread

### ISSUE THREAD

**Issue by pkpbynum** [software-mansion/react-native-reanimated#1641]
## Description
This is a bug/question, sorry it's so long...

TL;DR I'm using an IIFE and a closure on the UI thread to create a mutable object, but that object becomes immutable as soon as its accessed from `useAnimatedGestureHandler`. Why is that?

I'm rendering a list of children, each of which has a shared value. I want those shared values to be accessible by the parent. I thought I'd be clever and use an IIFE and a closure to create an object on the UI thread without using a shared value--so that its mutable--and use a worklet that gets this object wherever I need it. It looks like this:

```javascript
const getState = useRef(
    (() => {
      'worklet'
      const mutableObj = {
        offsets: {},
      }
      return () => {
        'worklet'
        return mutableObj
      }
    })(),
  ).current
```

In the onLayout for each child, I assign the shared value (`offset`):
```javascript
onLayout={() => {
          'worklet'
          getState().offsets[id] = offset
        }}
```

This works really well, and makes accessing/controlling the children from the parent very easy. The problem comes when I add `useAnimatedGestureHandler.` As soon as I try to access this state from within the handler, the onLayout no longer works. It looks like the object become immutable. Why is that?

I expect the answer here to be that it's complicated and not possible haha, but I'm wondering if this pattern is allowed at all? I also see a 'shared collection' feature on the horizon--would it cover this use case?

The main advantages this method has over using one sharedValue in the parent are:
1. I avoid any `JSON.parse(JSON.stringify(...))` shenanigans
2. I can independently animate each child

Am I way out of line with this?

## Steps To Reproduce

This is pure JS, so you can just copy/paste the below code if you have a project running. It's also under the "Closure" tab of [this GitHub repo](https://github.com/pkpbynum/reanimated-2-testing), which is the built off of the playground environment provided by the docs (below screenshot is what it looks like). There are 2 lines in the below code that are commented out. Uncommenting them makes the state immutable.

<img width="200" alt="Screen Shot 2021-01-16 at 6 46 53 PM" src="https://user-images.githubusercontent.com/34200453/104827406-36bb1380-582b-11eb-8d46-1c29472e54ab.png">

### Expected behavior
`getState()` in `useAnimatedGestureHandler` returns an object with each child's offset values. Something like:

```javascript
{
  "offsets": {
    "red": {
      "value": 0
    },
    "orange": {
      "value": 50
    },
    "blue": {
      "value": 100
    },
    "yellow": {
      "value": 150
    },
  }
}
```

### Actual behavior
The object becomes immutable, breaking the onLayout worklet. `getState()` actually returns something like:

```javascript
{
  "offsets": { }
}
```

## Snack or minimal code example
```javascript
import React, {useRef} from 'react'
import {View, Button} from 'react-native'
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  useAnimatedGestureHandler,
} from 'react-native-reanimated'
import ScreenWrapper from '../components/ScreenWrapper'
import {PanGestureHandler} from 'react-native-gesture-handler'

const ids = ['red', 'orange', 'blue', 'yellow']

const Child = ({id, index, getState}) => {
  const offset = useSharedValue(0)

  const gestureHandler = useAnimatedGestureHandler({
    onActive: (e) => {
      // const state = getState()
      // console.log(state)
      offset.value = e.translationY
    },
  })

  const animatedStyle = useAnimatedStyle(() => ({
    position: 'absolute',
    width: 50,
    height: 50,
    backgroundColor: id,
    transform: [{translateY: offset.value}, {translateX: index * 50}],
  }))

  return (
    <PanGestureHandler onGestureEvent={gestureHandler}>
      <Animated.View
        onLayout={() => {
          'worklet'
          getState().offsets[id] = offset
          console.log(getState())
        }}
        style={animatedStyle}
      />
    </PanGestureHandler>
  )
}

export default function Test() {
  const x = useSharedValue(0)
  const getState = useRef(
    (() => {
      'worklet'
      const mutableObj = {
        offsets: {},
      }
      return () => {
        'worklet'
        return mutableObj
      }
    })(),
  ).current

  const onPress = () => {
    'worklet'
    const state = getState()
    state[x.value] = x
    x.value = 10 + x.value
    console.log(state)
  }

  return (
    <ScreenWrapper>
      <Button onPress={onPress} title="Mutate Object" />
      <View>
        {ids.map((id, i) => (
          <Child key={id} id={id} index={i} getState={getState} />
        ))}
      </View>
    </ScreenWrapper>
  )
}

```

## Package versions

- React: 16.13.1
- React Native: 0.63.4
- React Native Reanimated: 2.0.0-rc.2
- NodeJS: 14.15.3
- React Native Gesture Handler: 1.9.0
